### DISCUSSION TITLE
deploying-asp-net-core-applications-to-kubernetes-part-6-adding-health-checks-with-liveness-readiness-and-startup-probes/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#17] [Announcements]
# deploying-asp-net-core-applications-to-kubernetes-part-6-adding-health-checks-with-liveness-readiness-and-startup-probes/

In this post I discuss the health check probes in Kubernetes, how to specify them in Helm and ASP.NET Core, and issues to watch out for.

https://andrewlock.net/deploying-asp-net-core-applications-to-kubernetes-part-6-adding-health-checks-with-liveness-readiness-and-startup-probes/

**Comment by AntonSmolkov** [comment]
Thanks for the good article!
> If you have multiple containers in a pod, then if any of the containers fail their liveness probes then the whole pod is killed and restarted.

This one seems to be incorrect. Individual containers are restarted separately.

**Reply by not-an-ai** [reply]
For reference:
> The kubelet uses liveness probes to know when to restart a container.

https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/

**Comment by MNF** [comment]
>>what people are checking their readiness checks so please let me know in the comments!  
“readiness probes …to deal with situations where the application is temporarily unavailable"
https://blog.devgenius.io/understanding-kubernetes-probes-5daaff67599a

You can point readiness probe to the same dump endpoint as livenessProbe, bit with **shorter** tolerance. 
E.g. if  pod is not responding to readiness probe after 3 attempts , we can stop new traffic for it.
If we set livenessProbe to try up to 10 attempts( by setting higher failureThreshold), we will wait more time before restart to give the pod a chance to recover.

In many cases(if expected startup time is less than  time of readiness timeout)   creation of **smart** startup probe may be  redundant, because readiness probe will not allow to have pod available, if starting of application is not completed/failed

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at October 06 2020, 06:14, in reply to [Toby J](https://disqus.com/by/irontoby/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/17#discussioncomment-6059560)</em>

---

<p>I think we're probably going to fundamentally (and maybe philosophically)  disagree on this one, but I'll give you my thoughts on it - I could be wrong of course, it's just what we've seen :)</p><p>&gt; Readiness checks don't cause cascading failures in the way you describe; Kubernetes won't kill the pod, it will just stop sending it traffic, which is exactly what you want if a dependency fails.</p><p>I guess it really depends on what the failure is. If it's a single pod, then yes, absolutely, taking it out of service and letting other pods handle it works well.</p><p>But what if the upstream dependency _is_ down. So none of the pods in the service can contact it? To downstream dependencies, if they can't see your service, then they go down too, and there's your cascading failure, with a single, original, failure as the source.</p><p>&gt; I don't see any situation where a pod has a failed dependency, and you would still want to send it traffic. For that reason it seems that readiness checks should absolutely check their dependencies.</p><p>I can think of lots :) Redis cache goes down? It's a dependency that I want to check at Startup but it doesn't stop me functioning. Maybe the "read replica" for a database is available, so I can service that traffic, but can't handle updates?</p><p>That all seems like resiliency stuff that needs to be built into your code, rather than at the Kubernetes level.</p><p>To be fair though, I can see both ways working. I just get scared by the "your whole app dies due to cascading failure" scenario that can happen (because it has to me, once 🤦‍♂️) It's always possible I was just holding it wrong to make that happen :)</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at October 06 2020, 07:34, in reply to [Toby J](https://disqus.com/by/irontoby/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/17#discussioncomment-6059563)</em>

---

<p>Absolutely, thanks for engaging :)</p><p>So the interesting thing to me is that final paragraph:</p><p></p><blockquote>So a "fine-grained" health check like Database read-write wouldn't be a "readiness" check if my app can stiff function without it, but a smarter "aggregate" check which returns Healthy if read-write is available, Degraded if read-only replica is available, or Unhealthy if none available would indeed be a readiness check in that scenario.</blockquote><p></p><p>Given that Kubernetes doesn't have the concept of "degraded", that means that you can essentially assume the db read-write check here is not in the readiness check (as it doesn't affect the outcome). Essentially what I'm doing here is taking that to its logical conclusion, and removing all those "don't matter" checks, and you're left with a simple liveness check! :)</p><p>Again, I think the difference in our attitude stems from what you think is most likely, and what your application can handle:</p><p>a) If a dependency (e.g. the database goes down), can the service still handle traffic, even if it's very degraded. If not, is this service a dependency of another service, where going offline would knock that one out too?<br>b) If a pod temporarily loses connection, can we handle that, or should we take the pod out of circulation.</p><p>In this post I made a bit of an error describing the situation - I talked about a "service" losing a connection, taking down the app. As you point out, a single pod losing connection wouldn't do that - it's only if all the pods in the service did. However, that could happen if, for example, the dependency (e.g. db) has a config error or drops off briefly for example.</p><p>Your comment has made me think I should put more faith in the readiness checks tbh... I think I just got so burned that one time that I'm too afraid to try it again 😉</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at October 19 2020, 10:00, in reply to [Blackclaws](https://disqus.com/by/blackclaws/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/17#discussioncomment-6059566)</em>

---

<p>I think it goes without saying that your database should be instrumented so that you know when it's gone down, without needing an application to indirectly tell you! :)</p><p>There's also another interesting point here - take your statement :</p><p></p><blockquote>If you lose a dependency that doesn't affect whether you can still service requests in general </blockquote><p></p><p>What if I have a cache so I can still serve <i>some</i> endpoints, even when the DB is down, but will fail for all requests that do the need the DB. Should we take the pods out of circulation? There's obviously nuance here, but this has the potential of causing a larger outage than necessary.</p><p>If we mark as degraded, then that means we don't use the check in a readiness check, essentially my suggestion. The fact other health check scrapers can use that information is good, but doesn't change my logic here :)</p><p>Thanks!</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at December 16 2020, 09:21, in reply to [swisshero](https://disqus.com/by/SwissHero/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/17#discussioncomment-6059572)</em>

---

<p>Yes, that is the approach we use, using the aspnetcore approach  :) The consistency is appealing. The node that runs the pods will cache the image layers yes, but as the aspnetcore layer is based on the .NET Core layer, you get caching in both cases, so not really an argument there :)</p>

[a comment by an automated account is not shown]

**Comment by AntonSmolkov** [comment]
>I suggest you use smart health checks for your startup probe, primarily to catch configuration errors.

BTW why don't just crash in case of configuration/dependency error? The result will be the same - application will be restarted.

**Comment by ericbl** [comment]
IMHO, the simplest way to handle the 'readiness' probe is to use the same as the 'startup' probe.
That's the way suggested by Microsoft documentation in https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/health-checks

Another use case that come in my mind for setting the probe to NOT ready is when any controller is returning the http status 429 / too many requests.
